src/components/Identity/createComponent.js

```diff
@@ -17,135 +17,135 @@ const addEcidToPayload = (payload, ecid) => {
+        }
       },
-      // Waiting for opt-in because we'll be reading the ECID from a cookie
-      // TO-DOCUMENT: We wait for ECID before trigger any events.
-      onBeforeDataCollection({ payload }) {
-        return optIn.whenOptedIn().then(() => {
-          let promise;
-
-          if (!hasIdentityCookie()) {
-            const ecidToMigrate =
-              idMigrationEnabled && migration.getEcidFromLegacyCookies();
-
-            if (ecidToMigrate) {
-              // We don't have an identity cookie, but we do have an ECID
-              // from a legacy cookie that we can explicitly provide
-              // to the server, which is sufficient until the identity cookie
-              // gets set.
-              addEcidToPayload(payload, ecidToMigrate);
-            } else if (deferredForEcid) {
-              // We don't have an identity cookie, but the first request has
-              // been sent to get it. We must wait for the response to the first
-              // request to come back and a cookie set before we can let this
-              // request go out.
-              logger.log("Delaying request while retrieving ECID from server.");
-              promise = deferredForEcid.promise.then(() => {
-                logger.log("Resuming previously delayed request.");
-              });
-            } else {
-              // We don't have an identity cookie and no request has gone out
-              // to get it. We'll let this request go out to fetch the cookie,
-              // but we'll set up a promise so that future requests can
-              // know when the cookie has been set. We don't let additional
-              // requests to go out in the meantime because a new ECID would
-              // be minted for each request (each request would be seen as a
-              // new visitor).
-              deferredForEcid = defer();
-              payload.expectResponse();
-              // If third-party cookies are enabled by the customer and
-              // supported by the browser, we will send the request to a
-              // a third-party domain that allows for more accurate
-              // identification of the user through use of a third-party cookie.
-              if (
-                config.thirdPartyCookiesEnabled &&
-                areThirdPartyCookiesSupportedByDefault(getBrowser(window))
-              ) {
-                payload.useIdThirdPartyDomain();
-              }
-            }
-          }
-
-          customerIds.addToPayload(payload);
-          return promise;
-        });
+      onBeforeDataCollectionRequest({ payload }) {
+        return accommodateIdentityOnRequest(payload);
+      },
+      onBeforeConsentRequest({ payload }) {
+        return accommodateIdentityOnRequest(payload);
       },
-
-      // Waiting for opt-in because we'll be reading the ECID from a cookie
       onResponse({ response }) {
-        return optIn.whenOptedIn().then(() => {
```

Comment:
This whole function body was wrapped in `optIn.whenOptedIn().then()`. I decided to push the logic that waits for consent closer to the actual thing that needs consent. In this case, we need consent when creating the legacy cookie and we need consent to process ID syncs, so I pushed the logic that waits for consent down into `migration.createLegacyCookie` and `processIdSyncs`. That prevents us from waiting for consent unnecessarily (like if we don't need to write a legacy cookie or process ID syncs) and helps ensure that we will wait for consent whenever we need to (for example, if some other module in the system started calling `migration.createLegacyCookie` or `processIdSyncs`).

Reply:
@jfkhoury are you suggesting adding a comment like "If you do anything privacy-sensitive within this lifecycle callback, be sure to call optIn.whenOptedIn first"? If so, it seems like that comment would apply to pretty much the whole Alloy codebase and not just this area of code. In other words, respecting privacy seems like something engineers need to keep in mind regardless of which part of Alloy they're working on and therefore a comment in this particular area of code doesn't really address the larger need to be privacy-conscious. Do you agree or do you still feel like we still need a comment here specifically?